content/thin-provisioning-redesign.md

```diff
@@ -0,0 +1,137 @@
+the two, but puts a rather strict limit on snapshot capacity. For users that use Stratis for
+critical applications or the root filesystem, this mode prevents certain failure cases that
+can be challenging to recover from.
+
+When overprovisioning is enabled, we have also introduced a new API signal to notify the user
+when physical storage has been fully allocated. This does not necessarily mean that the pool
+has run out of space but serves as a warning to the user that once the remaining free space
+fills up, Stratis has no space left to extend to. This gives users time to provide more storage
+from which to allocate space before reaching a failure case.
+
+Reliability
+===========
+For a while, we've gotten bug reports about the reliability of filesystem extension. In certain
+cases, Stratis was not able to handle filesystem extension smoothly or at all. Between the
+[per-pool locking] and the thin provisioning redesign, we have now resolved some of the
+previous issues with filesystem extension. The approach we've taken attacks the problem from
+a few different angles.
+
+Earlier filesystem extension
+----------------------------
+Stratis used to wait until several gigabytes were left to extend the filesystem. If Stratis didn't
+resize the filesystem quickly enough, the filesystem would run out of space before the extension
+could complete. While this would eventually resolve itself once the filesystem was extended,
+it would cause some unnecessary IO errors. We now extend the filesystem at 50% usage
+to ensure that users always have a large buffer of free space available for even very IO-heavy
+usage patterns.
+
+Parallelized filesystem extension operations
+--------------------------------------------
+Stratis could previously only iterate sequentially through pools. Now stratisd can handle filesystem
+extension on two separate pools in parallel, reducing the latency between the point where
+high usage is detected and the extension operation being performed.
+
+Periodic checks for filesystem usage
+------------------------------------
+Checking filesystem usage used to be a devicemapper event-dependent operation. This led to
+some problems around filesystem extension. A devicemapper event would be generated periodically
+as the filesystem filled up, but if the filesystem failed to extend a few times,
+devicemapper events would no longer be generated once the pool filled up and users would be
+left with a filesystem that couldn't be extended. We've removed our dependency on devicemapper
+events for filesystem operations and use devicemapper events for pool handling exclusively.
+In place, we run periodic checks in the background on filesystems to ensure that even if
+filesystem extension fails multiple times, once the filesystem is ready to be extended,
+stratisd can perform the operation the background, never leaving users in a state where their
```

Comment:
^the background^in the background

Also, "never" might be a little optimistic.